### ISSUE TITLE
Android : Creating video from images can't be sent anywhere (whatsapp, fb etc.)

### ISSUE THREAD

**Issue by raghavsatyadev** [jcodec/jcodec#364]
I already followed all the previous issues and sample code given in here. None of it seems to work. I am getting images from gallery or camera and then scale them according to proper resolution then adding them to video. Please help me ... 

I have posted the code here : https://gist.github.com/raghavsatyadev/8ee54f25206422720db98c2d0cd59d26

**Comment by svitvitskiy** [comment]
I'll take a look

**Comment by gekco** [comment]
same issue. but it works everywhere except whatsapp. Whatsapp crashes on sending such video. https://drive.google.com/open?id=0B4mBJIZp1C0Wc3RhcnRlcl9maWxl

**Comment by gekco** [comment]
@raghavsatyadev any luck?

**Comment by raghavsatyadev** [comment]
Nope. Never made any try after creating the issue, I had to use FFmpeg

**Comment by jadebeer** [comment]
Same problem in Android. Have to use the option in Whatsapp to send it as a gif. It sends in messenger and Facebook messenger, so it may be a Whatsapp issue with H264 ??? But seeing how pervasive Whatsapp use is, it would be great if this can be fixed.

**Comment by svitvitskiy** [comment]
Let me make sure I understand. So the video is produced, and it is playable on PC (Mac, Linux) but when you upload it to Whatsapp, the result is corrupted?

**Comment by jadebeer** [comment]
![Screenshot_20200906-222423](https://user-images.githubusercontent.com/9635419/92351334-3dbd6280-f090-11ea-8c90-efa93520aad2.jpg)

Yes, the video clip is fine but whatsapp doesn't like it. You can share it as gif on Whatsapp but not as an mp4/video.

**Comment by zhuker** [comment]
@jadebeer can you post the video file created and a code snippet?

**Comment by jadebeer** [comment]
@zhuker 

Hope this helps.

It doesn't allow me to attach the mp4 file. Please download it here: https://drive.google.com/file/d/1tUEJgAssZf2LOHLtswK040Bha-fw3-T1/view?usp=sharing


Part of a JobIntentService...

`SeekableByteChannel out = null;

    try {
      out = writableFileChannel(filePath);
      AndroidSequenceEncoder encoder = new AndroidSequenceEncoder(out, Rational.R(FRAMES_PER_SEC, 1));
      //SequenceEncoder encoder2 =
      //        new SequenceEncoder(out, Rational.R(FRAMES_PER_SEC, 1),
      //                Format.MOV, Codec.H264, null);
      Bundle bundleSteps = new Bundle();
      bundleSteps.putInt(STEPS_INFORM, FRAMES_PER_SEC * SEC_PER_IMG * uris.length);
      mResultReceiver.send(STEPS_MAX, bundleSteps);

      for (int index = 0; index < uris.length; index++) {
        if (TextUtils.isEmpty(uris[index])) {
          continue;
        }

        bmpUri = Uri.parse(uris[index]);

        images = getBitmapSetFromStorage(this, bmpUri);

        // Encode the image
        for (int i = 0; i < (FRAMES_PER_SEC * SEC_PER_IMG); i++) {
          encoder.encodeImage(images[1]);
          //encoder2.encodeNativeFrame(BitmapUtil.fromBitmap(images[1]));
          Bundle bundleUpdate = new Bundle();
          bundleUpdate.putInt(PROGRESS_INDEX, (FRAMES_PER_SEC * SEC_PER_IMG * index) + i + 1);
          mResultReceiver.send(PROGRESS_UPDATE, bundleUpdate);
        }
      }

      // Finalize the encoding, i.e. clear the buffers, write the header, etc.
      encoder.finish();
      //encoder2.finish();
    } catch (FileNotFoundException e) {
      e.printStackTrace();
    } catch (IOException e) {
      e.printStackTrace();
    } finally {
      NIOUtils.closeQuietly(out);
    }`

**Comment by zhuker** [comment]
@svitvitskiy where is our relocate `moov` atom to the beginning of the file code? i ran `ffmpeg -movflags faststart` and whatsapp shared it without problem

@jadebeer this looks like an issue specific to whatsapp, other apps share the video just fine, that said a workaround is to move `moov` atom to the beginning of the file, i'll try to find the code and let you know

**Comment by jadebeer** [comment]
Awesome thank you and much appreciated @zhuker I'll wait for your info.

**Comment by svitvitskiy** [comment]
@zhuker org.jcodec.movtool.WebOptimize
If this was just due to location of the MOOV then we could add a call to WebOptimize to org.jcodec.api.transcode.SinkImpl#finish so that to avoid confusion for the users.

**Comment by zhuker** [comment]
@jadebeer see https://github.com/jcodec/jcodec/blob/master/src/main/java/org/jcodec/movtool/WebOptimize.java let me know if it works

**Comment by jadebeer** [comment]
@zhuker I applied these changes but it made no difference. Unfortunately the produced file could not be shared through Whatsapp.
